### PR TITLE
Add interpreter for CollectivePermuteOp

### LINKED ISSUE #1124: Add interpreter for CollectivePermuteOp
Following the [testing guidelines](https://github.com/openxla/stablehlo/blob/main/docs/interpreter_checklist.md)

### PR DESCRIPTION
**Pull request by ghpvnist**
We have the following constraints in the spec:

```
(I1) `operand`: tensor.
(I2) `source_target_pairs`: 2-dimensional tensor constant of type `si64`.
(I3) `channel_id`: constant of type `si64`.
(C1) `dim(source_target_pairs, 1) = 2`.
(C2) `is_unique(source_target_pairs[:, 0])`.
(C3) `is_unique(source_target_pairs[:, 1])`.
(C4) `0 <= source_target_pairs < N`, where N is defined as:
* `num_replicas` if `cross_replica` is used.
* `num_partitions` if `cross_partition` is used.
(C5) `type(result) = type(operand)`.
```

These constraints will be comprehensively covered by the following tests:

```
I1: a) `operand` is not a tensor. (Covered by ODS).
I2: a) `source_target_pairs` is not a 2-dimensional tensor constant of type `si64`.
I3: a) `channel_id` is not a constant of type `si64`. (Covered by ODS).
C1: a) `dim(source_target_pairs, 1) != 2`.
C2: a) `is_unique(source_target_pairs[:, 0]) = false`.
C3: a) `is_unique(source_target_pairs[:, 1]) = false`.
C4: a) `source_target_pairs < 0`.
C4: b) `source_target_pairs >= N`, where `N` is defined as:
* `num_replicas` if `cross_replica` is used.
* `num_partitions` if `cross_partition` is used.
C5: a) `type(result) != type(operand)`.
```

If we drop the "Covered by ODS" pieces, this will leave us with the following test cases:

```
I2a: `source_target_pairs` is not a 2-dimensional tensor constant of type `si64`.
C1a: `dim(source_target_pairs, 1) != 2`.
C2a: `is_unique(source_target_pairs[:, 0]) = false`.
C3a: `is_unique(source_target_pairs[:, 1]) = false`.
C4a: `source_target_pairs < 0`.
C4b: `source_target_pairs >= N`, where `N` is defined as:
* `num_replicas` if `cross_replica` is used.
* `num_partitions` if `cross_partition` is used.
C5a: `type(result) != type(operand)`.
```

Notes:
  * C4b verification is infeasible since `num_replicas` and `num_partitions` are not known statically at the moment (see #650).

closes #1124

### CODE REVIEW

File: stablehlo/reference/ProcessGrid.h [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+  bool operator<(const ProcessId &other) const {
+    return std::pair<uint32_t, uint32_t>{replicaId, partitionId} <
+           std::pair<uint32_t, uint32_t>{other.replicaId, other.partitionId};
+  }
+
+  bool operator==(const ProcessId &other) const {
+    return std::pair<uint32_t, uint32_t>{replicaId, partitionId} ==
+           std::pair<uint32_t, uint32_t>{other.replicaId, other.partitionId};
+  }
```
Comment (burmako) [inline stablehlo/reference/ProcessGrid.h:48]: Let's move the implementations of the operations into the .cpp file, to make it easier to understand the APIs in this header.

File: stablehlo/reference/ProcessGrid.cpp [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  }
+  return processGroups;
+}
+
+SmallVector<std::pair<ProcessId, Tensor>> ProcessGrid::rendezvous(
+    ProcessGroup processGroup, int64_t channelId, ProcessId &processId,
+    const Tensor &operand) {
+  {
+    std::unique_lock<std::mutex> lk(
+        resourceLockMap_[{processGroup, channelId}]);
```
Comment (burmako) [inline stablehlo/reference/ProcessGrid.cpp:97]: From https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly: "Avoid abbreviations unless they are well known". I don't think we're gaining much by saving two characters at the cost of reducing readability.

File: stablehlo/reference/ProcessGrid.cpp [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    channels_[{processGroup, channelId}].push_back({processId, operand});
+  }
+
+  resourceConditionMap_[{processGroup, channelId}].notify_all();
```
Comment (burmako) [inline stablehlo/reference/ProcessGrid.cpp:101]: Why is this needed?

File: stablehlo/reference/ProcessGrid.cpp [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+                     processGroup.size();
+            }))
+      llvm::report_fatal_error("rendezvous timed out");
+
+    std::sort(channels_[{processGroup, channelId}].begin(),
+              channels_[{processGroup, channelId}].end(),
+              [](std::pair<ProcessId, Tensor> &lhs,
+                 std::pair<ProcessId, Tensor> &rhs) {
+                return lhs.first < rhs.first;
+              });
```
Comment (burmako) [inline stablehlo/reference/ProcessGrid.cpp:121]: I don't think the sort piece is needed at the moment. And in the future it could be hidden as an implementation detail of `RendezvousResult`.

File: stablehlo/reference/ProcessGrid.cpp (hunk 2 above)
Comment (burmako) [inline stablehlo/reference/ProcessGrid.cpp:97]: Should we use `std::lock_guard` here, given that we don't really use the additional functionality provided by `std::unique_lock`?

File: stablehlo/reference/ProcessGrid.h [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+  ProcessGroups crossReplica(SmallVector<SmallVector<uint32_t>> replicaGroups);
+
+  /// Cross-partition communication strategy.
+  ProcessGroups crossPartition(
+      SmallVector<SmallVector<uint32_t>> partitionGroups);
+
+  /// Synchronizes running processes in the `ProcessGroup` and returns the
+  /// collected resources from all partitipating processes in the process group.
+  SmallVector<std::pair<ProcessId, Tensor>> rendezvous(
+      ProcessGroup processGroup, int64_t channelId, ProcessId &processId,
```
Comment (burmako) [inline stablehlo/reference/ProcessGrid.h:79]: What do you think about defining `using ChannelId = int64_t` in this file, and then using that definition both here and in Ops.h/cpp.

File: stablehlo/reference/ProcessGrid.cpp (hunk 2 above)
Comment (burmako) [inline stablehlo/reference/ProcessGrid.cpp:97]: Let's refactor the key into something like `std::pair<ProcessGroup, ChannelId> channelKey(processGroup, channelId);`? It's used quite a few times in this method.

File: stablehlo/reference/ProcessGrid.h [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+#include <utility>
+
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/Support/LLVM.h"
+#include "stablehlo/reference/Tensor.h"
+
+namespace mlir {
+namespace stablehlo {
+
+// Object to represent StableHLO `process_id`.
```
Comment (burmako) [inline stablehlo/reference/ProcessGrid.h:32]: Simplify to "StableHLO `process_id`", similarly to "StableHLO `replica_id`" below? (Same for the comment on `ProcessGroup`).

File: stablehlo/reference/ProcessGrid.h [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+  }
+
+  bool operator==(const ProcessId &other) const {
+    return std::pair<uint32_t, uint32_t>{replicaId, partitionId} ==
+           std::pair<uint32_t, uint32_t>{other.replicaId, other.partitionId};
+  }
+};
+
+// Object to represent StableHLO process group.
+struct ProcessGroup : SmallVector<ProcessId> {};
```
Comment (burmako) [inline stablehlo/reference/ProcessGrid.h:52]: https://llvm.org/docs/CodingStandards.html#use-of-class-and-struct-keywords: "struct should be used when all members are declared public". Given that not all members of this program element are public, `class` is a better fit.

File: stablehlo/reference/ProcessGrid.h [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+  std::map<std::pair<ProcessGroup, int64_t>, std::mutex> resourceLockMap_;
+
+  /// Synchronization primitive used in `rendezvous` function.
+  std::map<std::pair<ProcessGroup, int64_t>, std::condition_variable>
+      resourceConditionMap_;
+
+  /// Internal mapping of StableHLO `channel_id` to its value.
+  std::map<std::pair<ProcessGroup, int64_t>,
+           SmallVector<std::pair<ProcessId, Tensor>>>
+      channels_;
```
Comment (burmako) [inline stablehlo/reference/ProcessGrid.h:99]:
```C++
  /// Internal storage used to implement `rendezvous`.
  /// Each call to `rendezvous`, i.e. each combination `processGroup` and
  /// `channelId`, has its own key in the map.
  /// Within the implementation of `rendezvous`, the value corresponding to
  /// this key is gradually populated with tensors arriving from different
  /// processes in the process group.
  std::map<std::pair<ProcessGroup, ChannelId>, RendezvousResult> channels_;

  /// Synchronization primitive used to manage concurrent access to `channels_`.
  std::map<std::pair<ProcessGroup, ChannelId>, std::mutex> channelLocks_;

  /// Synchronization primitive used to manage concurrent access to `channels_`.
  std::map<std::pair<ProcessGroup, ChannelId>, std::condition_variable>
      channelConditions_;
```

Feedback:
  1) It's unclear what's a value of a channel_id, so I elaborated on that.
  2) Dropped the `Map` prefixes, because I don't think much is gained by signaling in the name that this is a map - we can kind of get this already from the signature on the definition site and from the called methods on use sites.
  3) Renamed the `resource` prefix to clearly associate these maps with `channels_`.

File: stablehlo/reference/ProcessGrid.h [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+
+  /// Cross-partition communication strategy.
+  ProcessGroups crossPartition(
+      SmallVector<SmallVector<uint32_t>> partitionGroups);
+
+  /// Synchronizes running processes in the `ProcessGroup` and returns the
+  /// collected resources from all partitipating processes in the process group.
+  SmallVector<std::pair<ProcessId, Tensor>> rendezvous(
+      ProcessGroup processGroup, int64_t channelId, const Tensor &operand);
+};
```
Comment (burmako) [inline stablehlo/reference/ProcessGrid.h:120]: Let's split `class Process` into a separate file (Process.h/cpp)? This file is getting pretty intense, so modularizing it would be beneficial. (`ProcessId` would need to stay here, and that's fine imo).

**Review by burmako — CHANGES_REQUESTED** [review]

File: stablehlo/reference/ProcessGrid.cpp [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+  return processGroups;
+}
+
+SmallVector<std::pair<ProcessId, Tensor>> ProcessGrid::rendezvous(
+    ProcessGroup processGroup, int64_t channelId, ProcessId &processId,
+    const Tensor &operand) {
+  {
+    std::unique_lock<std::mutex> lk(
+        resourceLockMap_[{processGroup, channelId}]);
+    channels_[{processGroup, channelId}].push_back({processId, operand});
```
[the remark this answers is not in the document]
Comment (ghpvnist) [reply stablehlo/reference/ProcessGrid.cpp]: Updated.

File: stablehlo/reference/ProcessGrid.h [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+  bool operator==(const ProcessId &other) const {
+    return std::pair<uint32_t, uint32_t>{replicaId, partitionId} ==
+           std::pair<uint32_t, uint32_t>{other.replicaId, other.partitionId};
+  }
+};
+
+// Object to represent StableHLO process group.
+struct ProcessGroup : SmallVector<ProcessId> {};
+
+// Object to represent StableHLO process groups.
```
[the remark this answers is not in the document]
Comment (ghpvnist) [reply stablehlo/reference/ProcessGrid.h]: [StableHLO process groups](https://github.com/openxla/stablehlo/blob/main/docs/spec.md#collective-ops) is defined here.

File: stablehlo/reference/ProcessGrid.h (hunk 7 above)
Comment (ghpvnist) [reply stablehlo/reference/ProcessGrid.h]: Actually, why not just make all members public and let both `ProcessGroup` and `ProcessGroups` be a `struct`. Both are essentially an alias of `SmallVector`.

File: stablehlo/reference/Ops.cpp [hunk 12]
```diff
@@ -800,6 +820,37 @@ Tensor evalClzOp(const Tensor &operand, ShapedType resultType) {
...
+  if (!process)
+    llvm::report_fatal_error(
+        "collective_permute is only supported when run via "
+        "interpreter.run_parallel");
+
+  ProcessGroups processGroups;
+  if (channelId <= 0) processGroups = process->crossReplica(sourceTargetPairs);
+  if (channelId > 0) processGroups = process->crossPartition(sourceTargetPairs);
+
+  std::optional<Tensor> result;
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:835]: Given that Tensor supports the notion of empty tensors, we don't need to represent optional tensors via `std::optional`. Just `Tensor` would suffice. The only thing missing is this:

```C++
class Tensor {
  ...
  explicit operator bool() const { return (bool)impl_; }
  bool operator!() const { return !impl_; }
  ...
}
```

File: stablehlo/reference/ProcessGrid.h [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+  Tensor lookup(ProcessId processId);
+
+  /// Inserts `tensor` into the map `result_` using the key `processId`.
+  void insert(ProcessId processId, Tensor tensor);
+
+  /// Erases all elements in the map `result_`.
+  void clear();
+
+  /// Returns the size of `result_`.
+  size_t size();
```
Comment (burmako) [inline stablehlo/reference/ProcessGrid.h:69]: Let's write comments in terms of the public API, not private fields.

```C++
/// Represents a result of a `ProcessGrid::rendezvous` where multiple processes
/// synchronize at a barrier and contribute a Tensor each.
/// This class is pretty much a map from ProcessId to Tensor, with the 
/// map-like API.
class RendezvousResult {
  /// Iterates through the map `result_` and returns the tensor associated with
  /// the key `processId`.
  /// If key is not found, return an empty `Tensor`.
  Tensor lookup(ProcessId processId);

  /// Inserts `tensor` into the map using the key `processId`.
  void insert(ProcessId processId, Tensor tensor);

  /// Erases all elements in the map.
  void clear();

  /// Returns the number of elements in the map.
  size_t size();
}
```

File: stablehlo/reference/ProcessGrid.h [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+
+ private:
+  ///
+  std::map<ProcessId, Tensor> result_;
+};
+
+/// Class to model a process grid.
```
Comment (burmako) [inline stablehlo/reference/ProcessGrid.h:76]: "/// StableHLO process grid" similarly to above comments?

File: stablehlo/reference/ProcessGrid.h [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+class ProcessGrid {
+ public:
+  /// \name Constructors
+  /// @{
+  ProcessGrid(uint32_t numReplicas, uint32_t numPartitions);
+  /// @}
+
+  /// StableHLO `cross_replcia`.
```
Comment (burmako) [inline stablehlo/reference/ProcessGrid.h:84]: "/// StableHLO `cross_replica` communication strategy"? Same below.

File: stablehlo/reference/ProcessGrid.h [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+  ProcessGroups crossPartition(
+      SmallVector<SmallVector<uint32_t>> partitionGroups);
+
+  /// Each participating process in the `processGroup` appends its data
+  /// `operand` to the `channels_` map using the pair (processGroup, channelId)
+  /// as a key. `channels_` is cleared before any process adds its data.
+  /// `rendezvous` then returns `RendezvousResult` containing a mapping of
+  /// `processId` to its data `operand` once all participating processes have
+  /// successfully added their data. Throws an error after a timeout when
+  /// synchronization deadlocks.
```
Comment (burmako) [inline stablehlo/reference/ProcessGrid.h:97]: Same as above - let's write this comment in terms of a public API.

**Review by burmako — CHANGES_REQUESTED** [review]

File: stablehlo/reference/ProcessGrid.h [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+
+// StableHLO `process_id`.
+struct ProcessId {
+  /// StableHLO `replica_id`.
+  uint32_t replicaId;
+
+  /// StableHLO `partition_id`.
+  uint32_t partitionId;
+
+  bool operator<(const ProcessId &other) const;
```
[the remark this answers is not in the document]
Comment (ghpvnist) [reply stablehlo/reference/ProcessGrid.h:42]: It is still needed to sort the keys of `std::map<std::pair<ProcessGroup, ChannelId>, RendezvousResult> channels_`. We could use `std::unordered_map`, but this requires a hash function instead. We will soon need to use this operator to sort `RendezvousResult` anyways hence keeping this change.

[the remark this answers is not in the document]
Comment (burmako) [reply stablehlo/reference/ProcessGrid.h:42]: "We will soon need to use this operator to sort RendezvousResult anyways hence keeping this change". I think this operator should come with an update to the spec, because this piece is underspecced at the moment. Let's defer to a later PR.

"It is still needed to sort the keys". Hmm good point. Let's switch to a different container then. Let's use llvm::DenseMap to follow LLVM conventions.

[the remark this answers is not in the document]
Comment (burmako) [reply stablehlo/reference/ProcessGrid.h:42]: "because this piece is underspecced at the moment". Actually, I was wrong - the spec does talk about how `process_groups` are computed exactly. But I still think we shouldn't have `operator<`, since this can be an implementation detail of `RendezvousResult`.

[the remark this answers is not in the document]
Comment (ghpvnist) [reply stablehlo/reference/ProcessGrid.h:46]: >  Let's switch to a different container then. Let's use llvm::DenseMap to follow LLVM conventions.

I looked into `std::map`, `std::unordered_map`, and `llvm::DenseMap`, and of the three, `std::map` seemed to have the most straightforward, minimal implementation. Since we define how `process_groups` are computed for each communication strategy, we could define `operator<` on the `ProcessGroup`, but this seems like an anti-pattern. I've added a comment to justify the need.

File: stablehlo/reference/ProcessGrid.h (hunk 16 above)
Comment (burmako) [reply stablehlo/reference/ProcessGrid.h:104]: The current wording is still formulated in terms of the implementation details which involve an internal map with an internal key - something that users of this API shouldn't need to know about.

```C++
  /// Synchronize a StableHLO process with the `processId` with other StableHLO
  /// processes in the `processGroup` using a `channelId`.
  ///
  /// A call to this method represents a barrier, i.e. it blocks the calling
  /// OS thread until all StableHLO processes from the `processGroup` call this 
  /// method with the same `channelId`. If the calling OS thread doesn't 
  /// correspond to the StableHLO process with `processId`, the behavior is 
  /// undefined.
  ///
  /// If any of the StableHLO processes from `processGroup` fails to arrive
  /// at the barrier within 3 seconds, the `rendezvous` fails with a fatal 
  /// error. This is to make sure that errors in underlying StableHLO programs 
  /// or bugs in the StableHLO interpreter don't deadlock the entire 
  /// interpreter.
  ///
  /// At the barrier, each StableHLO process contributes a tensor, and these
  /// tensors are accumulated in `RendezvousResult` which is returned to all
  /// callers once the barrier has been reached by all StableHLO processes.
```

**Review by burmako — CHANGES_REQUESTED** [review]

File: stablehlo/reference/ProcessGrid.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (ghpvnist) [reply stablehlo/reference/ProcessGrid.h:46]: I opened a [ticket](https://github.com/openxla/stablehlo/issues/1720) to address this in the future.

**Review by burmako — APPROVED** [review]
